### ISSUE TITLE
InvalidOperationException thrown in GetBody() when using byte[] body

### ISSUE THREAD

**Issue by mcmellawatt** [Azure/amqpnetlite#232]
Some of our tests are failing with the latest AMQP .Net Lite version 2.0.0, and upon further inspection it appears to be due to some bad assumptions in the GetBody<T>() extension method for the Message class.  I believe this method used to be in the Message class but has been moved to an extension method with version 2.0.0.  I have found a workaround but I want to check if you think this is a bug or we are using the library incorrectly.  I've created some sample code to reproduce the problem:

```
using Amqp;
using Amqp.Framing;

namespace ReproAmqpBug
{
    class Program
    {
        static void Main(string[] args)
        {
            var payload = new byte[] { 1, 2, 3, 4, 5 };

            var message = new Message(payload)
            {
                Properties = new Properties
                {
                    Subject = "testsubject",
                    ContentEncoding = "testencoding",
                    ContentType = "testtype"
                }
            };

            var sentMessage = Message.Decode(message.Encode());

            var receivedPayload = sentMessage.GetBody<byte[]>();
        }
    }
}
```

We use "Message.Decode(message.Encode())" as a means of simulating sending the message over the wire in our tests.  You'll see at the last line that an InvalidOperationException is thrown using version 2.0.0 of the library, but not in previous versions of the library (we are using 1.2.2 specifically).  At a glance, it looks like the GetBinaryOffset() code is just grabbing the first byte from the buffer and assuming it is the format code byte.  However, in my example above, the first byte is occupied by other data (namely the message properties I believe).

The workaround here is very straight forward - I just change the last line to:

`var receivedPayload = sentMessage.Body as byte[];`

In fact, this appears to be the desired way to access the message body, judging by your example applications.  I thought I'd just bring the GetBody() issue in case it is a bug, and also see if there is a strong reason to use message.Body instead of message.GetBody<T>().

**Comment by mcmellawatt** [comment]
Thanks for the quick fix Xin!  For now, do you think our workaround valid?

`var receivedPayload = sentMessage.Body as byte[];`

instead of

var receivedPayload = sentMessage.GetBody<byte[]>();

If not, any way we can push this fix into a release?

**Comment by mcmellawatt** [comment]
Thanks for the quick fix Xin!  For now, do you think our workaround valid?

`var receivedPayload = sentMessage.Body as byte[];`

instead of

var receivedPayload = sentMessage.GetBody<byte[]>();

If not, any way we can push this fix into a release?

**Comment by mcmellawatt** [comment]
Thanks for the quick fix Xin!  For now, do you think our workaround valid?

`var receivedPayload = sentMessage.Body as byte[];`

instead of

var receivedPayload = sentMessage.GetBody<byte[]>();

If not, any way we can push this fix into a release?

**Comment by mcmellawatt** [comment]
Thanks for the quick fix Xin! For now, do you think our workaround valid?

var receivedPayload = sentMessage.Body as byte[];

instead of

var receivedPayload = sentMessage.GetBody<byte[]>();

If not, any way we can push this fix into a release?

**Comment by xinchen10** [comment]
@mcmellawatt it is a valid workaround. After the fix the two lines do exact the same thing. The GetBody method was refactored into an extension method due to the NuGet package separation on netstandard.

For large byte[] the recommended way to access the bytes is to call `message.GetBody<ByteBuffer>()`. The library creates a thin wrapper (ByteBuffer) of the underlying bytes without actually copying them. The ByteBuffer class is similar to ArraySegment<byte> (buffer.Buffer = segment.Array, buffer.Offset = segment.Offset, buffer.Length = segment.Count).

**Comment by mcmellawatt** [comment]
Yes actually we were using the `message.GetBody<ByteBuffer>()` for the very reason you just mentioned.  It ran into the same issue as the byte[] since both code paths use GetBinaryOffset() in the same way.  Once your fix is released, we will go back to using `message.GetBody<ByteBuffer>()`.  Thanks again for all your help!

**Comment by xinchen10** [comment]
Fix release in v2.1.0.
